testsuite/domain/src/test/java/org/jboss/as/test/integration/domain/management/cli/HierarchicalCompositionTestCase.java

```diff
@@ -0,0 +1,236 @@
+import org.junit.Before;
+import org.junit.Test;
+
+
+import java.io.IOException;
+
+import static org.junit.Assert.assertTrue;
+
+/**
+ * Test for logical hierarchical composition of Profiles and Socket Binding Groups.
+ *
+ * https://issues.jboss.org/browse/WFCORE-844
+ *
+ * @author Marek Kopecky <mkopecky@redhat.com>
+ */
+public class HierarchicalCompositionTestCase extends AbstractCliTestBase {
+
+    private static Logger log = Logger.getLogger(HierarchicalCompositionTestCase.class);
+
+    private static final String ORIGINAL_PROFILE = "default";
+
+    private static final String ROOT_PROFILE = "hierarchical-composition-test-case-root";
+
+    private static final String CHILD_PROFILE = "hierarchical-composition-test-case-default";
+
+    private static final String ROOT_SUBSYSTEM = "request-controller";
+
+    private static final String ORIGINAL_SERVER_GROUP = "main-server-group";
+
+    private static final String SERVER_GROUP = "test-group";
+
+    private static final String SERVER = "test-server";
+
+    private static final String HOST = "master";
+
+    private static final String NEW_SOCKET_BINDING_GROUP = "hierarchical-composition-test-case-binding-group";
+
+    private static final String NEW_SOCKET_BINDING = "hierarchical-composition-test-case-binding";
+
+    private static final int NEW_SOCKET_BINDING_PORT = 1234;
+
+    /**
+     * Original profile from server group
+     */
+    String initProfile;
+
+    /**
+     * Original socket binding group from server group
+     */
+    String initSocketBindingGroup;
+
+
+    @BeforeClass
+    public static void before() throws Exception {
+        CLITestSuite.createSupport(HierarchicalCompositionTestCase.class.getSimpleName());
+        AbstractCliTestBase.initCLI(DomainTestSupport.masterAddress);
+    }
+
+    @AfterClass
+    public static void after() throws Exception {
+        AbstractCliTestBase.closeCLI();
+        CLITestSuite.stopSupport();
+    }
+
+    /**
+     * Sends command line to CLI, validate and return output.
+     *
+     * @param line command line
+     * @return CLI output
+     */
+    private String cliRequest(String line, boolean successRequired) {
+        log.info(line);
+        cli.sendLine(line);
+        String output = cli.readOutput();
+        if (successRequired) {
+            assertTrue("CLI command \"" + line + " doesn't contain \"success\"", output.contains("success"));
+        }
+        return output;
+    }
+
+    /**
+     * Sends command line to CLI, return only result.
+     *
+     * @param line command line
+     * @return CLI output
+     */
+    private String cliGetResult(String line) {
+        log.info(line);
+        cli.sendLine(line);
+        String output = null;
+        try {
+            output = (String) cli.readAllAsOpResult().getResult();
+        } catch (IOException e) {
+            Assert.fail("Fail to get result of CLI command: " + line + "\n" + e.getMessage() + "\n" + e.getStackTrace());
```

Comment:
One small note here, but it doesn't need to be changed immediately. Instead of using `\n` we should use `System.lineSeparator()` so errors will look nice for Windows users as well. It's not a huge deal in a test really, just something to keep consistent on for good practice :)

Another option would be to use a string format like:
```java
String.format("Fail to get result of CLI command: %s%n%s%n%s", line, e.getMessage(), e.getStackTrace())
```